include/swift/AST/Identifier.h

```diff
@@ -203,7 +205,144 @@ namespace llvm {
 } // end namespace llvm
 
 namespace swift {
+
+/// Wrapper that may either be an Identifier or a special name
+/// (e.g. for subscripts)
+///
+/// The raw values of this enum are used in serialization by both
+/// BaseNameToEntitiesTable(Reader|Writer)Info and DeclTableInfo. For every
+/// reordering, deletion or insertion of any case, the major (resp. minor)
+/// version of both SwiftLookupTable AND ModuleFormat must thus be incremented.
+class DeclBaseName {
+public:
+  enum class Kind : uint8_t {
+    Normal = 0,
+    Subscript
+  };
   
+private:
+  /// In a special DeclName represenenting a subscript, this opaque pointer
+  /// is used as the data of the base name identifier.
+  /// This is an implementation detail that should never leak outside of
+  /// DeclName.
+  static void *SubscriptIdentifierData;
+
+  Identifier Ident;
+
+public:
+  DeclBaseName() : DeclBaseName(Identifier()) {}
+
+  DeclBaseName(Identifier I) : Ident(I) {}
+
+  static DeclBaseName createSubscript() {
+    return DeclBaseName(Identifier((const char *)SubscriptIdentifierData));
+  }
+
+  Kind getKind() const {
+    if (Ident.get() == SubscriptIdentifierData) {
+      return Kind::Subscript;
+    } else {
+      return Kind::Normal;
+    }
+  }
+
+  bool isSpecial() const { return getKind() != Kind::Normal; }
+
+  /// Return the identifier backing the name. Assumes that the name is not
+  /// special.
+  Identifier getIdentifier() const {
+    assert(!isSpecial() && "Cannot retrieve identifier from special names");
+    return Ident;
+  }
+
+  bool empty() const { return !isSpecial() && getIdentifier().empty(); }
+
+  bool isOperator() const {
+    return !isSpecial() && getIdentifier().isOperator();
+  }
+
+  bool isEditorPlaceholder() const {
+    return !isSpecial() && getIdentifier().isEditorPlaceholder();
+  }
+
+  /// A representation of the name to be displayed to users. May be ambiguous
+  /// between identifiers and special names
+  StringRef userFacingStr() const {
+    switch (getKind()) {
+    case Kind::Normal:
+      return getIdentifier().str();
+    case Kind::Subscript:
+      return "subscript";
+    }
+  }
+
+  int compare(DeclBaseName other) const {
+    return userFacingStr().compare(other.userFacingStr());
+  }
+
+  bool operator==(StringRef Str) const {
```

Comment:
I guess I was worried about {{== "subscript"}} (and whatever else we add), and maybe that's being too cautious.